enacts/wat_bal/agronomy.py

```diff
@@ -214,3 +214,105 @@ def soil_plant_water_balance(
+    api_poly=xr.DataArray(
+        [
+            [0.858, 0.0895, 0.0028],
+            [-1.14, 0.042, 0.0026],
+            [-2.34, 0.12, 0.0026],
+            [-2.36, 0.19, 0.0026],
+            [-2.78, 0.25, 0.0026],
+            [-3.17, 0.32, 0.0024],
+            [-4.21, 0.438, 0.0018],
+        ],
+        dims=["api_cat", "powers"],
+    ),
+    time_dim="T",
+):
+    """Computes Runoff using Antecedent Precipitation Index.
+    `runoff` is a polynomial of `daily_rain` of order 2.
+    Polynomial is chosen based on API categories.
+    Additionaly, `runoff` is 0 if it rains less or equal than `no_runoff` ,
+    and negative `runoff` is 0.
+    Parameters
+    ----------
+    daily_rain : DataArray
+        daily precipitation
+    no_runoff : DataArray, optional
+        `runoff` is 0 if `daily_rain` is leser or equal to `no_runoff`
+        (default `no_runoff` =12.5)
+    api_thresh : DataArray, optional
+        increasing daily API values along a dimension called api_cat
+        indicating the upper limit (inlcusive) to belong to an API category
+    api_poly : DataArray, optional
+        polynomial coefficients that must depend on a dimension called powers
+        of size 3 (the 3 powers of a polynomial of order 2),
+        and on a dimension api_cat of size one more than `api_thresh` 's size.
+        The polynomial used to compute the `runoff` is picked according to the categories
+        defined by the thresholds.
+    time_dim : str, optional
+        daily time dimension of `daily_rain` (default `time_dim` ="T").
+        
+    Returns
+    -------
+    runoff : DataArray
+        daily Runoff.
+    See Also
+    --------
+    api_sum
+    Notes
+    -----
+    For instance with the default parameters, if rain is greater or equal to 12.5
+    and API is 18, then runoff is -1.14 + 0.042*x 0.0026*(x**2)
+    where x is daily rain.
+    """
+    # Compute API
+    api = daily_rain.rolling(**{time_dim:7}).reduce(api_sum).dropna(time_dim)
+    runoff = xr.dot(
+        xr.dot(
+            api_poly,
+            xr.concat(
+                [np.power(daily_rain, 0), daily_rain, np.square(daily_rain)],
+                dim="powers",
+            ).where(daily_rain > no_runoff, 0),
+        ),
+        xr.concat(
+            [api <= api_thresh, ~np.isnan(api)],
+            dim="api_cat",
+        ).cumsum(dim="api_cat").where(lambda x: x <= 1, other=0),
+        dims="api_cat",
+    ).clip(min=0).rename("runoff")
+    runoff.attrs = dict(description="Runoff", units="mm")
+    return runoff
+
+
+def api_sum(a, axis=-1):
+    """Weighted-sum for Antecedent Precipitation Index for an array of length n
+    applies weights of 1/2 for last element and 1/(n-i-1) for all others
+    Parameters
+    ----------
+    a : array_like
+        elements to weight and sum
+    axis : int, optional
+        Axis along which the weight and sum are performed.
+        if None, applies to last axis.
```

Comment:
I am not sure why I did it that way back then. What I can think of now:

- I was under the impression that rolling.reduce applied to the grouped dimension created by rolling by design but I may be wrong. 
- api_sum is extremely specific so I think I aimed at factoring it out for ease of read and not meant it to be reused by anything else, and thus figured a simpler numpy think was sufficient

Reply:
> I was under the impression that rolling.reduce applied to the grouped dimension created by rolling

I checked the [docs](https://docs.xarray.dev/en/stable/generated/xarray.core.rolling.DataArrayRolling.html#xarray.core.rolling.DataArrayRolling). All the specific reduction functions like min, max, and mean say "Reduce this object's data windows by applying [min/max/mean] along its dimension," but rolling is different: it says "along some dimension(s)." By specifying axis=-1 you _might_ always be picking the right dimension, but it's not clear to me that that's the case. I would be more comfortable if you used `.reduce(api_sum, dim=time_dim)`.  I'm not sure if that will work directly with your numpy implementation of api_sum, or if api_sum will have to be rewritten with xarray.

Reply:
I've left this aside for now. The documentation's [examples](https://docs.xarray.dev/en/stable/generated/xarray.core.rolling.DataArrayRolling.reduce.html#xarray.core.rolling.DataArrayRolling.reduce) suggest it applies only to the rolled dimension/window.
Will get back to it once we are happy with the main function.

Reply:
There is now a dedicated api function as opposed to api_sum to feed an rolling calculation. The whole point was in case someone wanted to calculate runoff differently. The way it is, they now can by computing their own api, provide their own api tresholdds and polynomials.